### PR TITLE
replace cast string to fixed function in driver.cpp

### PR DESCRIPTION
**Pull request by AEsir777**
ToDo
- ~add more tests~
- ~fixed float conversion bug (probably just because the digits are full)~
- ~error message~

| | new | old |
|---|---|----|
|`FLOAT(20)` x 100000| 5194.5 | 7127.13200|

### CODE REVIEW

File: src/processor/operator/persistent/reader/csv/driver.cpp [hunk 1]
```diff
@@ -204,6 +204,52 @@ static void castStringToList(const char* input, uint64_t len, ValueVector* vecto
...
+    if (numElementsRead != numElementsInList) {
+        throw CopyException(StringUtils::string_format(
+            "Each fixed list should have fixed number of elements. Expected: {}, Actual: {}.",
+            numElementsInList, numElementsRead));
+    }
+}
+
+static void castStringToFixedList(const char* input, uint64_t len, ValueVector* vector,
+    uint64_t rowToAdd, const CSVReaderConfig& csvReaderConfig) {
+    // assertions
```
Comment (acquamarin) [inline src/processor/operator/persistent/reader/csv/driver.cpp:218]: Remove this trivial comment

File: src/processor/operator/persistent/reader/csv/driver.cpp [hunk 2]
```diff
@@ -204,6 +204,52 @@ static void castStringToList(const char* input, uint64_t len, ValueVector* vecto
...
+    assert(vector->dataType.getLogicalTypeID() == LogicalTypeID::FIXED_LIST);
+    auto childDataType = FixedListType::getChildType(&vector->dataType);
+    switch (childDataType->getLogicalTypeID()) {
+    case LogicalTypeID::INT64:
+    case LogicalTypeID::INT32:
+    case LogicalTypeID::INT16:
+    case LogicalTypeID::INT8:
+    case LogicalTypeID::UINT64:
```
Comment (acquamarin) [inline src/processor/operator/persistent/reader/csv/driver.cpp:226]: I think we only support fixedlist of int64,int32,int16, can you confirm with @Ashleyhx

File: src/processor/operator/persistent/reader/csv/driver.cpp [hunk 3]
```diff
@@ -204,6 +204,52 @@ static void castStringToList(const char* input, uint64_t len, ValueVector* vecto
...
+    case LogicalTypeID::UINT32:
+    case LogicalTypeID::UINT16:
+    case LogicalTypeID::UINT8:
+    case LogicalTypeID::DOUBLE:
+    case LogicalTypeID::FLOAT:
+        break;
+    default: {
+        throw CopyException("Unsupported data type " +
```
Comment (acquamarin) [inline src/processor/operator/persistent/reader/csv/driver.cpp:234]: Add a test for this exception

File: src/processor/operator/persistent/reader/csv/driver.cpp [hunk 4]
```diff
@@ -21,7 +21,7 @@ namespace processor {
 ParsingDriver::ParsingDriver(common::DataChunk& chunk) : chunk(chunk), rowEmpty(false) {}
 
 void copyStringToVector(ValueVector* vector, uint64_t rowToAdd, std::string_view strVal,
```
Comment (acquamarin) [inline src/processor/operator/persistent/reader/csv/driver.cpp:23]: I think curType is only used for fixedList right?
To make the interface and implementation cleaner, you can use var-size template and implement specialization for each dataType.

**Review by acquamarin — APPROVED** [review]

File: src/processor/operator/persistent/reader/csv/driver.cpp (hunk 2 above)
Comment (Ashleyhx) [reply src/processor/operator/persistent/reader/csv/driver.cpp:226]: yes i think so

File: src/processor/operator/persistent/reader/csv/driver.cpp [hunk 5]
```diff
@@ -522,6 +599,81 @@ void copyStringToVector(ValueVector* vector, uint64_t rowToAdd, std::string_view
...
+    function::doubleCast<float_t>(
+        strVal.data(), strVal.length(), value, LogicalType{LogicalTypeID::FLOAT});
+    vector->setValue(rowToAdd, value);
+}
+
+template<>
+void copyStringToVector(ValueVector* vector, uint64_t rowToAdd, std::string_view strVal,
+    const CSVReaderConfig& csvReaderConfig, double_t value) {
+    if (strVal.empty() || isNull(strVal)) {
+        vector->setNull(rowToAdd, true /* isNull */);
```
Comment (andyfengHKU) [inline src/processor/operator/persistent/reader/csv/driver.cpp:666]: Just make sure we add test to cover missing cases. Here and across the PR

File: src/processor/operator/persistent/reader/csv/driver.cpp [hunk 6]
```diff
@@ -522,6 +599,81 @@ void copyStringToVector(ValueVector* vector, uint64_t rowToAdd, std::string_view
...
+    }
+
+    function::simpleIntegerCast<int32_t>(
+        strVal.data(), strVal.length(), value, LogicalType{LogicalTypeID::INT32});
+    vector->setValue(rowToAdd, value);
+}
+
+template<>
+void copyStringToVector(ValueVector* vector, uint64_t rowToAdd, std::string_view strVal,
+    const CSVReaderConfig& csvReaderConfig, int16_t value) {
```
Comment (acquamarin) [inline src/processor/operator/persistent/reader/csv/driver.cpp:634]: The casting for numeric types are very similar: You can implement the casting for numeric types in the template function body, and use specialization for other types.  So you don't need to duplicate the code for numeric type casting.
The numeric type casting:
```
 if (strVal.empty() || isNull(strVal)) {
        vector->setNull(rowToAdd, true /* isNull */);
        return;
    } else {
        vector->setNull(rowToAdd, false /* isNull */);
    }

    function::simpleIntegerCast<T>(
        strVal.data(), strVal.length(), value, vector->getDataType());
    vector->setValue(rowToAdd, value);
```
Using this generic  templated numeric type casting function, you don't need to duplicate the common code for all numeric types.
Anyways, feel free to ping me if you have questions regarding to the refactor. Or you can leave it to me.

File: src/processor/operator/persistent/reader/csv/driver.cpp [hunk 7]
```diff
@@ -522,6 +599,81 @@ void copyStringToVector(ValueVector* vector, uint64_t rowToAdd, std::string_view
     }
 }
 
+template<>
+void copyStringToVector(ValueVector* vector, uint64_t rowToAdd, std::string_view strVal,
```
Comment (acquamarin) [inline src/processor/operator/persistent/reader/csv/driver.cpp:603]: Every copyStringToVector specialization has this common code block:
```
if (strVal.empty() || isNull(strVal)) {
        vector->setNull(rowToAdd, true /* isNull */);
        return;
    } else {
        vector->setNull(rowToAdd, false /* isNull */);
    }
```
The only difference is the logic to cast and set value in the vector. So you only need to template the cast and set value logic.

**Pull request merged**